### ISSUE TITLE
`devtools::test()` sources helper files in wrong environment

### ISSUE THREAD

**Issue by wch** [r-lib/devtools#2301]
tl;dr: `devtools::test()` sources helper files in the package environment on the search path, but they should be sourced in the package's namespace.

In a helper file in httpuv (https://github.com/rstudio/httpuv/blob/55c8699d3c811521618e9ce04cacad57c1e89d80/tests/testthat/helper-app.R#L1-L22), we do `library(curl)`, and then call a function from it. The code looks roughly like this:

```R
library(curl)

foo <- function() {
   multi_run()
}
```

When I run `devtools::test()` in new R session, the tests call `foo()`, but it can't find `multi_run()`

```R
> devtools::test()
Loading httpuv
Testing httpuv
✔ |  OK F W S | Context
✖ |   2 3     | basic [0.2 s]                                                                                                      
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Error (test-app.R:30:3): Basic functionality
Error: could not find function "multi_run"
Backtrace:
 1. fetch(local_url("/", s1$getPort())) test-app.R:30:2
 2. curl_fetch_async(url, handle = handle) helper-app.R:83:2
 3. poll() helper-app.R:17:2
```

However, `testthat:test_local()` works:

```R
> testthat::test_local()
✔ |  OK F W S | Context
✔ |  17       | basic [0.4 s]                                                                                                      
✔ |   1       | frame completion                                                                                                   
✔ |   5       | http-parse [0.3 s]                                                                                                 
✔ | 149       | static [2.0 s]                                                                                                     
✔ |   0     8 | traffic                                                                                                            
```

The problem with `devtools::test()` seems to be that the helper is sourced into a sibling of the httpuv package environment (note: this is not the httpuv namespace, but the package env on the search path), and the curl package env is added as a child of the httpuv package env, so it's not visible from the file.

I edited `helper-app.R` so that it prints out parent envs of the current environment and global environment:

```R
library(curl)

cat("Parents of current env ==============================\n")
print(pryr::parenvs())

cat("Parents of global env ==============================\n")
print(pryr::parenvs(globalenv(), all = T))
```

This shows why objects from curl aren't visible from the helper file when using `devtools::test()`:

```
> devtools::test()
Loading httpuv
Parents of current env ==============================
Registered S3 method overwritten by 'pryr':
  method      from
  print.bytes Rcpp
   label                            name               
1  <environment: package:httpuv>    "package:httpuv"   
2  <environment: package:testthat>  "package:testthat" 
3  <environment: package:stats>     "package:stats"    
4  <environment: package:graphics>  "package:graphics" 
5  <environment: package:grDevices> "package:grDevices"
6  <environment: package:utils>     "package:utils"    
7  <environment: package:datasets>  "package:datasets" 
8  <environment: 0x7ffd4b687f10>    "custom"           
9  <environment: package:methods>   "package:methods"  
10 <environment: 0x7ffd4a2aa400>    "Autoloads"        
11 <environment: base>              ""                 
12 <environment: R_EmptyEnv>        ""                 
Parents of global env ==============================
   label                            name               
1  <environment: R_GlobalEnv>       ""                 
2  <environment: package:promises>  "package:promises" 
3  <environment: package:curl>      "package:curl"     
4  <environment: package:httpuv>    "package:httpuv"   
5  <environment: package:testthat>  "package:testthat" 
6  <environment: package:stats>     "package:stats"    
7  <environment: package:graphics>  "package:graphics" 
8  <environment: package:grDevices> "package:grDevices"
9  <environment: package:utils>     "package:utils"    
10 <environment: package:datasets>  "package:datasets" 
11 <environment: 0x7ffd4b687f10>    "custom"           
12 <environment: package:methods>   "package:methods"  
13 <environment: 0x7ffd4a2aa400>    "Autoloads"        
14 <environment: base>              ""                 
15 <environment: R_EmptyEnv>        ""                 
```

However, with `testthat::test_local()`, it does work, apparently because the helpers are sourced in the httpuv namespace (not package env):

```R
> testthat::test_local()
Parents of current env ==============================
Registered S3 method overwritten by 'pryr':
  method      from
  print.bytes Rcpp
  label                           name            
1 <environment: namespace:httpuv> ""              
2 <environment: 0x7faae5475168>   "imports:httpuv"
3 <environment: namespace:base>   ""              
4 <environment: R_GlobalEnv>      ""              
Parents of global env ==============================
   label                            name               
1  <environment: R_GlobalEnv>       ""                 
2  <environment: package:promises>  "package:promises" 
3  <environment: package:curl>      "package:curl"     
4  <environment: 0x7faae4924ab0>    "devtools_shims"   
5  <environment: package:httpuv>    "package:httpuv"   
6  <environment: package:testthat>  "package:testthat" 
7  <environment: package:stats>     "package:stats"    
8  <environment: package:graphics>  "package:graphics" 
9  <environment: package:grDevices> "package:grDevices"
10 <environment: package:utils>     "package:utils"    
11 <environment: package:datasets>  "package:datasets" 
12 <environment: 0x7faae3cafd10>    "custom"           
13 <environment: package:methods>   "package:methods"  
14 <environment: 0x7faae52aa400>    "Autoloads"        
15 <environment: base>              ""                 
16 <environment: R_EmptyEnv>        ""                 
```

**Comment by jimhester** [comment]
The dev version of devtools does use `testthat::test_local()`, so this should already be fixed once it has been released on CRAN.
